url11fb0973cf_raichuman9_Swimmer526_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/fr/inria/optimization/cmaes/CMAEvolutionStrategy.java

```diff
@@ -0,0 +1,2704 @@
+                   s2 = s;
+                   g = c * e[i];
+                   h = c * p;
+                   r = math.hypot(p,e[i]);
+                   e[i+1] = s * r;
+                   s = e[i] / r;
+                   c = p / r;
+                   p = c * d[i] - s * g;
+                   d[i+1] = h + s * (c * g + s * d[i]);
+    
+                   // Accumulate transformation.
+    
+                   for (int k = 0; k < n; k++) {
+                      h = V[k][i+1];
+                      V[k][i+1] = s * V[k][i] + c * h;
+                      V[k][i] = c * V[k][i] - s * h;
+                   }
+                }
+                p = -s * s2 * c3 * el1 * e[l] / dl1;
+                e[l] = s * p;
+                d[l] = c * p;
+    
+                // Check for convergence.
+    
+             } while (Math.abs(e[l]) > eps*tst1);
+          }
+          d[l] = d[l] + f;
+          e[l] = 0.0;
+       }
+      
+       // Sort eigenvalues and corresponding vectors.
+    
+       for (int i = 0; i < n-1; i++) {
+          int k = i;
+          double p = d[i];
+          for (int j = i+1; j < n; j++) {
+             if (d[j] < p) { // NH find smallest k>i
+                k = j;
+                p = d[j];
+             }
+          }
+          if (k != i) {
+             d[k] = d[i]; // swap k and i 
+             d[i] = p;   
+             for (int j = 0; j < n; j++) {
+                p = V[j][i];
+                V[j][i] = V[j][k];
+                V[j][k] = p;
+             }
+          }
+       }
+    } // tql2
+
+    /** not really in use so far, just clones and copies
+     * 
+     * @param popx genotype
+     * @param popy phenotype, repaired
+     * @return popy
+     */
+    double[][] genoPhenoTransformation(double[][] popx, double[][] popy) {
+    	if (popy == null || popy == popx || popy.length != popx.length) 
+    		popy = new double[popx.length][];
+    	
+    	for (int i = 0; i < popy.length; ++i)
+    		popy[i] = genoPhenoTransformation(popx[i], popy[i]);
+    	
+    	return popy;
+    }
+    /** not really in use so far, just clones and copies
+     * 
+     * @param popx genotype
+     * @param popy phenotype, repaired
+     * @return popy
+     */
+    double[][] phenoGenoTransformation(double[][] popx, double[][] popy) {
+    	if (popy == null || popy == popx || popy.length != popx.length) 
+    		popy = new double[popx.length][];
+    	
+    	for (int i = 0; i < popy.length; ++i)
+    		popy[i] = phenoGenoTransformation(popx[i], popy[i]);
+    	
+    	return popy;
+    }
+
+    /** not really in use so far, just clones and copies
+     * 
+     * @param x genotype
+     * @param y phenotype
+     * @return y
+     */
+    double[] genoPhenoTransformation(double[] x, double[] y) {
+    	if (y == null || y == x || y.length != x.length) {
+    		y = x.clone();
+    		return y; // for now return an identical copy
+    	}
+    	for(int i = 0; i < N; ++i)
+    		y[i] = x[i];
+    	return y;		
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1437:1443, 1452:1458.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.